File: go/cmd/jsonic/args.go
// Copyright (c) 2020-2026 Richard Rodger, Oliver Sturm, and other
// contributors, MIT License

package main

import (
	"strings"

	jsonic "github.com/tabnas/jsonic/go"
	tabnas "github.com/tabnas/parser/go"
)

// cliArgs holds the parsed command line, matching the TypeScript `args`
// object.
type cliArgs struct {
	help    bool
	stdin   bool
	debug   bool
	sources []string
	files   []string
	options []string
	meta    []string
	plugins []string
}

// parseArgs replicates the TypeScript argument loop exactly: a leading "-"
// is STDIN, "--" stops flag parsing, the value-taking flags consume the
// next argument, and any other "-" prefixed token before "--" is ignored
// as an unknown flag (note: in the TS CLI it is pushed to sources, so we
// match that — an unknown flag becomes a source, which jsonic then parses;
// the --not-an-arg-so-ignored test relies on this being harmless when a
// real source follows). To match TS, an unrecognized "-" token is added to
// sources.
func parseArgs(argv []string) cliArgs {
	var args cliArgs
	acceptArgs := true

	for i := 0; i < len(argv); i++ {
		arg := argv[i]

		if acceptArgs && strings.HasPrefix(arg, "-") {
			switch {
			case arg == "-":
				args.stdin = true
			case arg == "--":
				acceptArgs = false
			case arg == "--file" || arg == "-f":
				if i+1 < len(argv) {
					i++
					args.files = append(args.files, argv[i])
				}
			case arg == "--option" || arg == "-o":
				if i+1 < len(argv) {
					i++
					args.options = append(args.options, argv[i])
				}
			case arg == "--meta" || arg == "-m":
				if i+1 < len(argv) {
					i++
					args.meta = append(args.meta, argv[i])
				}
			case arg == "--debug" || arg == "-d":
				args.debug = true
				args.meta = append(args.meta, "log=-1")
			case arg == "--help" || arg == "-h":
				args.help = true
			case arg == "--plugin" || arg == "-p":
				if i+1 < len(argv) {
					i++
					args.plugins = append(args.plugins, argv[i])
				}
			case arg == "--nice" || arg == "-n":
				args.options = append(args.options, "JSON.space=2")
			default:
				args.sources = append(args.sources, arg)
			}
		} else {
			args.sources = append(args.sources, arg)
		}
	}

	return args
}

// handleProps replicates the TypeScript handle_props: each "name=value"
// string sets a dotted-path property in a bag, with the value parsed by
// vanilla jsonic. Entries with an empty name or empty value are skipped
// (the bad-args tests: "", "=", "bad=" are all no-ops).
func handleProps(propvals []string) map[string]any {
	out := map[string]any{}
	for _, propval := range propvals {
		// split(/=/) in TS splits on every "=", but only pv[0] and pv[1]
		// are used; everything after the first "=" is discarded. SplitN
		// here keeps pv[1] as the remainder, but the TS code reads pv[1]
		// (the segment between the first and second "="). Match TS: take
		// the text up to the second "=".
		name, value, hadEq := cut(propval, "=")
		if !hadEq {
			continue
		}
		// pv[1] in TS is the segment up to the next "=".
		value, _, _ = cut(value, "=")
		if name == "" || value == "" {
			continue
		}
		v, err := jsonic.Parse(value)
		if err != nil {
			continue
		}
		setProp(out, name, v)
	}
	return out
}

// cut is strings.Cut (Go 1.18+), spelled out for clarity of intent.
func cut(s, sep string) (before, after string, found bool) {
	return strings.Cut(s, sep)
}

// setProp sets a dotted-path property in a nested map[string]any bag,
// creating intermediate maps as needed. Mirrors tabnas util.prop /
// the TS util.prop.
func setProp(bag map[string]any, path string, val any) {
	parts := strings.Split(path, ".")
	cur := bag
	for i := 0; i < len(parts)-1; i++ {
		k := parts[i]
		next, ok := cur[k].(map[string]any)
		if !ok {
			next = map[string]any{}
			cur[k] = next
		}
		cur = next
	}
	cur[parts[len(parts)-1]] = val
}

// engineOptions returns the subset of the option bag that configures the
// engine, dropping the CLI-only JSON and plugin keys (which the engine
// would not understand). MapToOptions ignores unknown keys, but dropping
// them keeps intent clear and avoids any future surprise.
func engineOptions(bag map[string]any) map[string]any {
	out := map[string]any{}
	for k, v := range bag {
		if k == "JSON" || k == "plugin" {
			continue
		}
		out[k] = v
	}
	return out
}

// pluginOptionMap returns the options.plugin.<name> sub-bag, or nil.
func pluginOptionMap(bag map[string]any, name string) map[string]any {
	pl, ok := bag["plugin"].(map[string]any)
	if !ok {
		return nil
	}
	po, ok := pl[name].(map[string]any)
	if !ok {
		return nil
	}
	return po
}

// lookupPlugin resolves a plugin reference name against the registry,
// trying the bare name first and then a "@tabnas/"-stripped tail, matching
// the spirit of the TS handle_plugins fallback (require(name) then
// require('@tabnas/'+name)). Registry keys are plain reference names
// (e.g. "csv"); a path-like reference is reduced to its base name.
func lookupPlugin(registry map[string]tabnas.Plugin, name string) (tabnas.Plugin, bool) {
	if registry == nil {
		return nil, false
	}
	if p, ok := registry[name]; ok {
		return p, true
	}
	// Try the base name of a path-like reference (e.g. "../test/p0" -> "p0").
	base := name
	if i := strings.LastIndexAny(base, "/\\"); i >= 0 {
		base = base[i+1:]
	}
	base = strings.TrimSuffix(base, ".js")
	if p, ok := registry[base]; ok {
		return p, true
	}
	// Try @tabnas/-prefixed reference stripped to its tail.
	if strings.HasPrefix(name, "@tabnas/") {
		tail := strings.TrimPrefix(name, "@tabnas/")
		if p, ok := registry[tail]; ok {
			return p, true
		}
	}
	return nil, false
}

// bagToMeta converts the meta bag into the engine's meta map. The engine's
// ParseMeta takes a map[string]any directly. The CLI-injected "log" key
// (from --debug) has no Go engine meaning and is harmless to pass through.
func bagToMeta(bag map[string]any) map[string]any {
	if len(bag) == 0 {
		return nil
	}
	return bag
}

File: rs/src/args.rs
/* Copyright (c) 2020-2026 Richard Rodger, Oliver Sturm, and other
contributors, MIT License */

//! Argument parsing and the dotted-path property bags behind `-o` and
//! `-m`. The port of `go/cli/args.go` and of the argument loop at the top
//! of `ts/src/jsonic-cli.ts`.

use serde_json::{Map, Value as Json};

/// The parsed command line, matching the TypeScript `args` object.
#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub(crate) struct CliArgs {
    pub help: bool,
    pub stdin: bool,
    pub debug: bool,
    pub sources: Vec<String>,
    pub files: Vec<String>,
    pub options: Vec<String>,
    pub meta: Vec<String>,
    pub plugins: Vec<String>,
}

/// Parse `argv` (the program name first, as `std::env::args` yields it).
///
/// This replicates the TypeScript loop exactly. A bare `-` means standard
/// input, `--` stops flag parsing, the value-taking flags consume the
/// following argument, and every other token, `-` prefixed or not, becomes
/// a source. Only the exact spellings are flags: `--option=JSON.space=2`
/// is not `--option`, so it falls through to the sources and is parsed as
/// relaxed JSON, which is what the canonical command does with it.
///
/// A value-taking flag in the LAST position consumes nothing rather than
/// taking an absent value, mirroring the TypeScript `aI + 1 < argv.length`
/// guard and Go's `i+1 < len(argv)`.
pub(crate) fn parse_args(argv: &[String]) -> CliArgs {
    let mut args = CliArgs::default();
    let mut accept_args = true;
    let mut i = 1;

    while i < argv.len() {
        let arg = argv[i].as_str();

        // The value of a value-taking flag, when one follows it.
        let value = |i: &mut usize| -> Option<String> {
            if *i + 1 < argv.len() {
                *i += 1;
                Some(argv[*i].clone())
            } else {
                None
            }
        };

        if accept_args && arg.starts_with('-') {
            match arg {
                "-" => args.stdin = true,
                "--" => accept_args = false,
                "--file" | "-f" => args.files.extend(value(&mut i)),
                "--option" | "-o" => args.options.extend(value(&mut i)),
                "--meta" | "-m" => args.meta.extend(value(&mut i)),
                "--debug" | "-d" => {
                    args.debug = true;
                    args.meta.push("log=-1".to_string());
                }
                "--help" | "-h" => args.help = true,
                "--plugin" | "-p" => args.plugins.extend(value(&mut i)),
                "--nice" | "-n" => args.options.push("JSON.space=2".to_string()),
                _ => args.sources.push(arg.to_string()),
            }
        } else {
            args.sources.push(arg.to_string());
        }

        i += 1;
    }

    args
}

/// Split a `name=value` argument the way the TypeScript
/// `propval.split(/=/)` does: `pv[0]` is the text before the first `=`,
/// and `pv[1]` the text between the first and the second. Everything
/// after that is discarded. `None` when there is no `=` at all, which
/// leaves `pv[1]` undefined and the property unset.
fn split_propval(propval: &str) -> Option<(&str, &str)> {
    let (name, rest) = propval.split_once('=')?;
    let value = rest.split_once('=').map_or(rest, |(head, _)| head);
    Some((name, value))
}

/// Build a dotted-path property bag from `name=value` arguments, with each
/// leaf value parsed by vanilla jsonic. The port of `handle_props`.
///
/// An entry with an empty name or an empty value is a no-op, which is what
/// makes `-o ""`, `-o "="` and `-o "bad="` harmless. A leaf value that
/// does not parse is an error, as it is in TypeScript, where `Jsonic(...)`
/// throws out of `run()`.
pub(crate) fn handle_props(propvals: &[String]) -> Result<Json, String> {
    let mut out = Map::new();

    for propval in propvals {
        let Some((name, value)) = split_propval(propval) else {
            continue;
        };
        if name.is_empty() || value.is_empty() {
            continue;
        }
        let parsed = tabnas_jsonic::parse(value).map_err(|error| error.to_string())?;
        set_prop(&mut out, name, to_json(&parsed));
    }

    Ok(Json::Object(out))
}

/// The engine's value as `serde_json`, with whole numbers narrowed to
/// integers.
///
/// [`tabnas::Value::to_json`] keeps every number a float, because that is
/// what the engine holds. An option document reader that wants a count
/// asks for an integer (`options.debug.maxlen` does), and a float-backed
/// `serde_json::Number` answers `as_u64` with `None` however round it is,
/// so `-o debug.maxlen=11` would be rejected as "not a non-negative
/// integer". Narrowing here is what makes the numeric options settable.
pub(crate) fn to_json(value: &tabnas::Value) -> Json {
    match value {
        tabnas::Value::Number(number) => number_to_json(*number),
        tabnas::Value::Array(items) => Json::Array(items.iter().map(to_json).collect()),
        tabnas::Value::ListRef(list) => Json::Array(list.value.iter().map(to_json).collect()),
        tabnas::Value::Object(entries) => Json::Object(
            entries
                .iter()
                .map(|(key, value)| (key.clone(), to_json(value)))
                .collect(),
        ),
        tabnas::Value::MapRef(map) => Json::Object(
            map.value
                .iter()
                .map(|(key, value)| (key.clone(), to_json(value)))
                .collect(),
        ),
        other => other.to_json(),
    }
}

/// A finite whole number as an integer, anything else as it stands.
fn number_to_json(number: f64) -> Json {
    if number.fract() == 0.0 && number.is_finite() {
        if number >= 0.0 && number <= u64::MAX as f64 {
            return Json::Number((number as u64).into());
        }
        if number >= i64::MIN as f64 && number < 0.0 {
            return Json::Number((number as i64).into());
        }
    }
    serde_json::Number::from_f64(number).map_or(Json::Null, Json::Number)
}

/// Set a dotted path in a nested object bag, building the intermediate
/// objects. The port of the engine's `util.prop`: an intermediate that is
/// not an object is replaced by one.
fn set_prop(bag: &mut Map<String, Json>, path: &str, value: Json) {
    let parts: Vec<&str> = path.split('.').collect();
    let (last, parents) = parts
        .split_last()
        .expect("split never yields an empty list");

    let mut cursor = bag;
    for part in parents {
        let next = cursor
            .entry(part.to_string())
            .or_insert_with(|| Json::Object(Map::new()));
        if !next.is_object() {
            *next = Json::Object(Map::new());
        }
        cursor = next.as_object_mut().expect("just made it an object");
    }
    cursor.insert(last.to_string(), value);
}

/// The sub-object at `key`, or `None` when the bag has no object there.
pub(crate) fn sub_object<'a>(bag: &'a Json, key: &str) -> Option<&'a Map<String, Json>> {
    bag.get(key).and_then(Json::as_object)
}

/// The bag with the CLI-only `JSON` key removed: what is left configures
/// the engine.
///
/// `plugin` STAYS. It is an engine option, and leaving it in the document
/// is what routes `-o plugin.<name>.<option>=<value>` to the plugin,
/// exactly as it does in TypeScript: the engine merges
/// `options.plugin[<plugin name>]` into the bag a plugin is installed
/// with, keyed by the PLUGIN'S OWN name rather than by the `-p`
/// reference. That is why `-p ../test/pa-qa.js -o plugin.paqa.q=3`
/// reaches the plugin at all.
pub(crate) fn engine_options(bag: &Json) -> Json {
    let mut out = Map::new();
    if let Some(map) = bag.as_object() {
        for (key, value) in map {
            if key == "JSON" {
                continue;
            }
            out.insert(key.clone(), value.clone());
        }
    }
    Json::Object(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn argv(list: &[&str]) -> Vec<String> {
        std::iter::once("jsonic")
            .chain(list.iter().copied())
            .map(String::from)
            .collect()
    }

    #[test]
    fn a_trailing_value_flag_consumes_nothing() {
        for flag in ["-o", "-m", "-f", "-p"] {
            let args = parse_args(&argv(&["a:1", flag]));
            assert_eq!(args.sources, vec!["a:1".to_string()]);
            assert!(args.options.is_empty());
            assert!(args.files.is_empty());
            assert!(args.plugins.is_empty());
        }
    }

    #[test]
    fn end_of_flags_turns_the_rest_into_sources() {
        let args = parse_args(&argv(&["--", "-h", "-n"]));
        assert!(!args.help);
        assert_eq!(args.sources, vec!["-h".to_string(), "-n".to_string()]);
    }

    #[test]
    fn an_equals_form_is_not_a_flag() {
        let args = parse_args(&argv(&["--option=JSON.space=2"]));
        assert!(args.options.is_empty());
        assert_eq!(args.sources, vec!["--option=JSON.space=2".to_string()]);
    }

    #[test]
    fn a_propval_reads_only_the_first_two_segments() {
        assert_eq!(split_propval("a=b=c"), Some(("a", "b")));
        assert_eq!(split_propval("a=b"), Some(("a", "b")));
        assert_eq!(split_propval("a="), Some(("a", "")));
        assert_eq!(split_propval("a"), None);
    }

    #[test]
    fn a_dotted_path_builds_intermediate_objects() {
        let bag = handle_props(&["a.b.c=1".to_string()]).expect("parses");
        assert_eq!(bag.to_string(), r#"{"a":{"b":{"c":1}}}"#);
    }

    #[test]
    fn a_whole_option_value_narrows_to_an_integer() {
        let bag = handle_props(&["debug.maxlen=11".to_string()]).expect("parses");
        assert_eq!(bag["debug"]["maxlen"].as_u64(), Some(11));
        let bag = handle_props(&["a=1.5".to_string()]).expect("parses");
        assert_eq!(bag["a"].as_f64(), Some(1.5));
    }
}
